File: lab7/Dictionary.java
// Austin Yen
// 5/27/15
// Dictionary.java
// Dictionary ADT, provides methods for use
// in creating a tree of keys and their values


public class Dictionary implements DictionaryInterface{


	// private types and functions ------------------------------------------------

	// NodeObj
	private class Node{
	   String key;
	   String value;
	   Node left;
	   Node right;
	   public Node(String k, String v){
		   key = k;
		   value = v;
		   left = null;
		   right = null;
	   }
	}
	private Node root;
	private int numPairs;



	// findKey()
	// returns the Node containing key k in the subtree rooted at R, or returns
	// null if no such Node exists
	Node findKey(Node R, String k){
	   if(R==null || k.equals(R.key)) 
	      return R;
	   if( k.compareTo(R.key) < 0) 
	      return findKey(R.left, k);
	   else // k.compareTo(R.key)>0
	      return findKey(R.right, k);
	}


	// findParent()
	// returns the parent of N in the subtree rooted at R, or returns null 
	// if N is equal to R. (pre: R != null)
	Node findParent(Node N, Node R){
	   Node P= null;
	   if( N!=R ){
	      P = R;
	      while( P.left!=N && P.right!=N ){
	         if(N.key.compareTo(P.key)<0)
	            P = P.left;
	         else
	            P = P.right;
	      }
	   }
	   return P;
	}

	// findLeftmost()
	// returns the leftmost Node in the subtree rooted at R, or null if R is null.
	Node findLeftmost(Node R){
	   Node L = R;
	   if( L!=null ) for( ; L.left!=null; L=L.left) ;
	   return L;
	}

	// printInOrder()
	// appends the (key, value) pairs belonging to the subtree rooted at R in order
	// of increasing keys to a stringbuffer pointed to by a.
	void printInOrder(Node R, StringBuffer a){
	   if( R!=null ){
	      printInOrder(R.left, a);
	      a.append(R.key+" "+R.value+"\n");
	      printInOrder(R.right, a);
	   }
	}

	// deleteAll()
	// deletes all Nodes in the subtree rooted at N.
	void deleteAll(Node N){
	   if( N!=null ){
	      deleteAll(N.left);
	      deleteAll(N.right);
	   }
	}

	// public functions -----------------------------------------------------------

	// Dictionary()
	// constructor for the Dictionary type
	public Dictionary(){
	   root = null;
	   numPairs = 0;
	}

	// isEmpty()
	// returns true if this Dictionary is empty, false otherwise
	// pre: none
	public boolean isEmpty(){
	   return(numPairs==0);
	}

	// size()
	// returns the number of entries in this Dictionary
	// pre: none
	public int size(){
	   return(numPairs);
	}

	// lookup()
	// returns value associated key, or null reference if no such key exists
	// pre: none
	public String lookup(String key){
	   Node N;
	   N = findKey(root, key);
	   if(N == null){
		   return null;
	   }
	   return N.value;
	}

	// insert()
	// inserts new (key,value) pair into this Dictionary
	// pre: lookup(key)==null
	public void insert(String key, String value) throws KeyCollisionException{
	   Node N, A, B;
	   if( lookup(key)!=null ){
	      throw new KeyCollisionException("cannot insert duplicate key");
	   }
	   N = new Node(key, value);
	   B = null;
	   A = root;
	   while( A!=null ){
	      B = A;
	      if( key.compareTo(A.key)<0 ) A = A.left;
	      else A = A.right;
	   }
	   if( B==null ) root = N;
	   else if( key.compareTo(B.key) < 0) B.left = N;
	   else B.right = N;
	   numPairs++;
	}

	// delete()
	// deletes pair with the given key
	// pre: lookup(key)!=null
	public void delete(String key) throws KeyNotFoundException{
	   Node N, P, S;
	   N = findKey(root, key);
	   if( N==null ){
	      throw new KeyNotFoundException("cannot delete nonexistent key");
	   }
	   if( N.left==null && N.right==null ){
	      if( N==root ){
	         root = null;
	      }else{
	         P = findParent(N, root);
	         if( P.right==N ) P.right = null;
	         else P.left = null;
	      }
	   }else if( N.right==null ){
	      if( N==root ){
	         root = N.left;
	      }else{
	         P = findParent(N, root);
	         if( P.right==N ) P.right = N.left;
	         else P.left = N.left;
	      }
	   }else if( N.left==null ){
	      if( N==root ){
	         root = N.right;
	      }else{
	         P = findParent(N, root);
	         if( P.right==N ) P.right = N.right;
	         else P.left = N.right;
	      }
	   }else{  // N.left!=null && N.right!=null
	      S = findLeftmost(N.right);
	      N.key = S.key;
	      N.value = S.value;
	      P = findParent(S, N);
	      if( P.right==S ) P.right = S.right;
	      else P.left = S.right;
	   }
	   numPairs--;
	}

	// makeEmpty()
	// pre: none
	public void makeEmpty(){
	   deleteAll(root);
	   root = null;
	   numPairs = 0;
	}

	// toString()
	// returns a String representation of this Dictionary
	// overrides Object's toString() method
	// pre: none
	public String toString(){
		StringBuffer dict = new StringBuffer("");
		printInOrder(root,dict);
		String dicti = dict.toString();
		return dicti;
	}


}

File: pa5/Dictionary.c
//-----------------------------------------------------------------------------
// Austin Yen
// 6/2/15
// Implementation file for Dictionary ADT
// Dictionary.c
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include"Dictionary.h"
const int tableSize = 101;

// private types --------------------------------------------------------------

// NodeObj
typedef struct NodeObj{
	char* key;
	char* value;
	struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* x, char* y) {
	Node N = malloc(sizeof(NodeObj));
	assert(N != NULL);
	N->value = y;
	N->key = x;
	N->next = NULL;
	return(N);
}


// freeNode()
// destructor for the Node type
void freeNode(Node* pN){
	if (pN != NULL && *pN != NULL){
		free(*pN);
		*pN = NULL;
	}
}


// DictionaryObj
typedef struct DictionaryObj{
	int numItems;
	Node *table;
} DictionaryObj;

// private helper function -------------------------------------------------

// rotate_left()
// rotate the bits in an unsigned int
unsigned int rotate_left(unsigned int value, int shift) {
	int sizeInBits = 8 * sizeof(unsigned int);
	shift = shift & (sizeInBits - 1);
	if (shift == 0)
		return value;
	return (value << shift) | (value >> (sizeInBits - shift));
}
// pre_hash()
// turn a string into an unsigned int
unsigned int pre_hash(char* input) {
	unsigned int result = 0xBAE86554;
	while (*input) {
		result ^= *input++;
		result = rotate_left(result, 5);
	}
	return result;
}
// hash()
// turns a string into an int in the range 0 to tableSize-1
int hash(char* key){
	return (pre_hash(key) % tableSize);
}



// findKey()
// returns a reference to the Node at position with the same key
Node findKey(Dictionary D, char* key){
	Node N = NULL;
	if (D->numItems == 0){
		return NULL;
	}
	int i;
	for (i = 0; i < tableSize; i++){
		if (D->table[i] != NULL){
			N = D->table[i];
			break;
		}
	}
	if (strcmp(N->key, key) == 0){
		return N;
	}
	while (N->next != NULL){
		if (strcmp(N->key, key) == 0){
			return N;
		}
		N = N->next;
	}
	
	for (int j = i+1; j < tableSize; j++){
		if (D->table[j] == NULL){
			continue;
		}
		else{
			N = D->table[j];
			if (strcmp(N->key, key) == 0){
				return N;
			}
			while (N->next != NULL){
				if (strcmp(N->key, key) == 0){
					return N;
				}
				N = N->next;
			}
		}
	}
	
	return NULL;
}



// public functions -----------------------------------------------------------





// Dictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void){
	Dictionary D = malloc(sizeof(DictionaryObj));
	assert(D != NULL);
	D->numItems = 0;
	D->table = calloc(tableSize, sizeof(NodeObj));
	assert(D->table != NULL);
	for (int i = 0; i < tableSize; i++) D->table[i] = NULL;
	return D;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD){
	if (pD != NULL && *pD != NULL){
		
		free((*pD)->table);
		if (!isEmpty(*pD)) makeEmpty(*pD);
		free(*pD);
		*pD = NULL;
		
	}
}

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){
	if (D == NULL){
		fprintf(stderr,
			"Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	return(D->numItems == 0);
}

// size()
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D){
	return D->numItems;
}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no 
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* k){
	if (findKey(D, k) == NULL){
		return NULL;
	}
	else{
		return findKey(D, k)->value;
	}
}

// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v){
	int index = hash(k);
	if (lookup(D, k) == NULL){
		Node N;
		if(D->table[index] == NULL){ // if there are no items in that index, one is added
			N = newNode(k, v);
			D->table[index] = N;

		}

		else{ // if there are items in the index, a new node is placed at the end of the linked list
			N = D->table[index];
			
			Node O = newNode(k, v);
			O->next = N;
			D->table[index] = O;

		}
		D->numItems++;
	}
	
	else{
		fprintf(stderr,
			"Dictionary Error: Cannot insert duplicate key\n");
		exit(EXIT_FAILURE);
	}
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char* k){
	if (lookup(D, k) == NULL){
		fprintf(stderr,
			"Dictionary Error: cannot delete non-existent key\n");
		exit(EXIT_FAILURE);
	}
	else{

		if (D->table[hash(k)] == findKey(D, k)){
			Node N = D->table[hash(k)];
			D->table[hash(k)] = D->table[hash(k)]->next;
			freeNode(&N);
		}
		else{
			Node N = D->table[hash(k)];
			int x = 1;
			while (x == 1){
				if (N->next == findKey(D, k)){
					Node P = N;
					N = P->next;
					P->next = N->next;
					freeNode(&N);
					break;
				}
				N = N->next;
			}

		}
		D->numItems--;
	}
}

// makeEmpty()
// re-sets D to the empty state.
// pre: none
void makeEmpty(Dictionary D){
	if (D == NULL){
		fprintf(stderr,
			"Dictionary Error: calling makeEmpty() on NULL Dictionary reference\n");
		exit(EXIT_FAILURE);
	}
	while (D->numItems>0){
		for (int i = 0; i < tableSize; i++){
			while (D->table[i] != NULL){
				delete(D, D->table[i]->key);
			}
		}
		
	}

}

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out
void printDictionary(FILE* out, Dictionary D){
	for (int i = 0; i < tableSize; i++){
		if (D->table[i] != NULL){
			for (Node N = D->table[i]; N != NULL; N = N->next){
				fprintf(out, "%s %s\n", N->key, N->value);
			}
		}
	}


}

